### PR TITLE
Highlight parts of fn in type errors

### PR DESCRIPTION
**Pull request by estebank**
When a type error arises between two fn items, fn pointers or tuples,
highlight only the differing parts of each.

Examples:

### CODE REVIEW

**Comment by jonas-schievink** [comment]
Nice!

File: src/librustc_errors/diagnostic.rs [hunk 1]
```diff
@@ -52,6 +52,13 @@ impl DiagnosticStyledString {
     pub fn push_highlighted<S: Into<String>>(&mut self, t: S) {
         self.0.push(StringPart::Highlighted(t.into()));
     }
+    pub fn push<S: Into<String>>(&mut self, t: S, highlight: bool) {
+        if highlight {
+            self.0.push(StringPart::Highlighted(t.into()));
```
Comment (Centril) [inline src/librustc_errors/diagnostic.rs]:
```suggestion
            self.push_highlighted(t);
```

File: src/librustc_errors/diagnostic.rs [hunk 2]
```diff
@@ -52,6 +52,13 @@ impl DiagnosticStyledString {
...
+        } else {
+            self.0.push(StringPart::Normal(t.into()));
```
Comment (Centril) [inline src/librustc_errors/diagnostic.rs]:
```suggestion
            self.push_normal(t);
```

File: src/librustc/infer/error_reporting/mod.rs [hunk 3]
```diff
@@ -868,6 +867,120 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+        //                                     ^^^^^^^^
+        let output1 = sig1.output();
+        let output2 = sig2.output();
+        let (x1, x2) = self.cmp(output1, output2);
+        if !output1.is_unit() {
+            values.0.push_normal(" -> ".to_string());
+            (values.0).0.extend(x1.0);
+        }
+        if !output2.is_unit() {
+            values.1.push_normal(" -> ".to_string());
```
Comment (Centril) [inline src/librustc/infer/error_reporting/mod.rs]:
```suggestion
            values.1.push_normal(" -> ");
```
The API accepts `Into<String>`, so this should work? (applies throughout the function and in cmp below)

File: src/librustc/infer/error_reporting/mod.rs [hunk 4]
```diff
@@ -1124,6 +1237,64 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                    (values.0).0.extend(x1.0);
+                    (values.1).0.extend(x2.0);
+                    self.push_comma(&mut values.0, &mut values.1, len, i);
+                }
+                if len == 1 { // Keep the output for single element tuples as `(ty,)`.
+                    values.0.push_normal(",".to_string());
+                    values.1.push_normal(",".to_string());
+                }
+                values.0.push_normal(")".to_string());
+                values.1.push_normal(")".to_string());
```
Comment (Centril) [inline src/librustc/infer/error_reporting/mod.rs]: there's more :)

Comment (estebank) [reply src/librustc/infer/error_reporting/mod.rs]: done :)

**Review by davidtwco — APPROVED** [review]
LGTM, this is a really great improvement! Like @Centril has mentioned, `cmp_fn_sig` feels quite repetitive, but I can't think of a nice way to improve it. Feel free to r=me if you're happy with this.

**Comment by estebank** [comment]
@bors r=davidtwco

@davidtwco @Centril, I will make a follow up PR to split out `fn cmp` and will try to look for ways to cleanly deduplicate the code.

**Pull request merged**